libvast/vast/co_result.hpp

```diff
@@ -0,0 +1,1048 @@
+  }
+
+  /// Returns a pointer to the underlying typed response promise, or nullptr if
+  /// that does not yet exist. Use with caution.
+  caf::typed_response_promise<value_type>*
+  unsafe_try_get_response_promise() noexcept {
+    return std::get_if<caf::typed_response_promise<value_type>>(&storage_);
+  }
+
+private:
+  /// An internal storage container for the various possible result types.
+  /// - `std::monostate` indicates that no value was returned yet.
+  /// - `caf::typed_response_promise<value_type>` indicates that the coroutine
+  ///   was suspended, leading to the result being made available later.
+  /// - `value_type` indicates that a we returned a value.
+  /// - `caf::error` indicates that a we returned an error.
+  /// - `caf::delegated<value_type>` indicates that we delivered a result early,
+  ///   or moved responsibility for delivering to an external context.
+  std::variant<std::monostate, caf::typed_response_promise<value_type>,
+               value_type, caf::error, caf::delegated<value_type>>
+    storage_;
+};
+
+} // namespace detail
+
+/// A coroutine-enabling wrapper around `caf::result<T>`.
+///
+/// This is best explained using a tony table:
+///
+/// BEFORE
+///
+///     [self](T&...) -> caf::result<R> {
+///       A();
+///       self->request(d1, t, xs...).then(
+///         [](response_type& value) {
+///           B(value);
+///         },
+///         [](caf::error& err) {
+///           C(err);
+///         });
+///       D();
+///       return r;
+///     }
+///
+/// AFTER
+///
+///     co_lift([self](T...) -> co_result<R> {
+///       A();
+///       auto task = co_request_then(*self, d1, t, xs...);
+///       D();
+///       auto tag = co_await co_deliver(r);
+///       if (auto value = co_await task)
+///         B(*value);
+///       else
+///         C(value.error());
+///       co_return tag;
+///     })
+///
+/// The main benefit is immediately obvious: We've got rid of callback hell.
+///
+/// There exist a few pitfalls with this that might be easy to run into:
+/// - Using `self->request`, `self->make_response_promise`, and `self->delegate`
+///   within a `co_result` coroutine leads to undefined behavior and is
+///   inherently unsafe. Use `co_request_then(self, ...)`,
+///   `co_request_await(self, ...)`, `co_make_response_promise<T>(self)`, and
+///   `co_delegate(self, ...)` instead.
+/// - After awaiting a response handle, `self->current_sender()` might return an
+///   unexpected value because the resumed coroutine behaves as if it was called
+///   from the response handle continuation.
+/// - To deliver values early, i.e., on first suspend, use `co_deliver(...)`,
+///   which returns a tag type that later must be returned from the coroutine.
+/// - CAF supports catching exceptions via `set_exception_handler`. Throwing
+///   exceptions from a `co_result<T>` coroutine will always bypass CAF's
+///   exception handler and call `std::terminate()`.
+///
```

Comment:
Add a new paragraph that explains the interaction of the promise type with the CAF runtime, and how lifetime are managed exactly.